sdk/core/core/src/az_json_builder.c

```diff
@@ -2,220 +2,627 @@
 // SPDX-License-Identifier: MIT
 
 #include "az_hex_private.h"
+#include "az_json_private.h"
 #include "az_json_string_private.h"
 #include "az_span_private.h"
 #include <az_json.h>
 #include <az_span_internal.h>
 
 #include <_az_cfg.h>
 
+enum
+{
+  // Max size for an already escaped string value (~ half of INT_MAX)
+  _az_MAX_ESCAPED_STRING_SIZE = 1000000000,
+
+  // In the worst case, an ASCII character represented as a single UTF-8 byte could expand 6x when
+  // escaped.
+  // For example: '+' becomes '\u0043'
+  // Escaping surrogate pairs (represented by 3 or 4 UTF-8 bytes) would expand to 12 bytes (which is
+  // still <= 6x).
+  _az_MAX_EXPANSION_FACTOR_WHILE_ESCAPING = 6,
+
+  _az_MAX_UNESCAPED_STRING_SIZE
+  = _az_MAX_ESCAPED_STRING_SIZE / _az_MAX_EXPANSION_FACTOR_WHILE_ESCAPING, // 166_666_666 bytes
+};
+
 static AZ_NODISCARD az_span _get_remaining_span(az_json_builder* json_builder)
 {
-  return az_span_slice_to_end(json_builder->_internal.json, json_builder->_internal.length);
+  _az_PRECONDITION_NOT_NULL(json_builder);
+
+  return az_span_slice_to_end(
+      json_builder->_internal.destination_buffer, json_builder->_internal.bytes_written);
 }
 
-AZ_NODISCARD static az_result az_json_builder_append_str(az_json_builder* self, az_span value)
+// TODO: Make this a precondition?
```

Comment:
In a normal, deployed, running app, should this ever false? And, if so, what corrective action could be taken at runtime? If not possible, then I'd make this a precondition check to not waste code/speed in a deployed/running app